runtime/fastly/host-api/host_api.cpp

```diff
@@ -10,26 +10,92 @@
 #include <algorithm>
 #include <arpa/inet.h>
 
+#include <time.h>
+
 using api::FastlyResult;
 using fastly::FastlyAPIError;
+using host_api::MonotonicClock;
 using host_api::Result;
 
+#define NEVER_HANDLE 0xFFFFFFFE
+
 size_t api::AsyncTask::select(std::vector<api::AsyncTask *> *tasks) {
   size_t tasks_len = tasks->size();
-  fastly_compute_at_edge_async_io_handle_t *handles =
-      new fastly_compute_at_edge_async_io_handle_t[tasks_len];
-  for (int i = 0; i < tasks_len; i++) {
-    handles[i] = tasks->at(i)->id();
+  std::vector<fastly_compute_at_edge_async_io_handle_t> handles;
+  handles.reserve(tasks_len);
+  uint64_t now = 0;
+  uint64_t soonest_deadline = 0;
+  size_t soonest_deadline_idx = -1;
+  for (size_t idx = 0; idx < tasks_len; ++idx) {
+    auto task = tasks->at(idx);
+    uint64_t deadline = task->deadline();
+    // Select for completed task deadlines before performing the task select host call.
+    if (deadline > 0) {
+      MOZ_ASSERT(task->id() == NEVER_HANDLE);
+      if (now == 0) {
+        now = MonotonicClock::now();
+        MOZ_ASSERT(now > 0);
+      }
+      if (deadline <= now) {
+        return idx;
+      }
+      if (soonest_deadline == 0 || deadline < soonest_deadline) {
+        soonest_deadline = deadline;
+        soonest_deadline_idx = idx;
+      }
+    } else {
+      uint32_t handle = task->id();
+      // Timer task handles are skipped and never passed to the host.
+      MOZ_ASSERT(handle != NEVER_HANDLE);
+      handles.push_back(handle);
+    }
+  }
+
+  // When there are no async tasks, sleep until the deadline
+  if (handles.size() == 0) {
+    MOZ_ASSERT(soonest_deadline > 0);
+    while (soonest_deadline > now) {
+      uint64_t duration = soonest_deadline - now;
+      timespec req{.tv_sec = static_cast<time_t>(duration / 1000000000),
+                   .tv_nsec = static_cast<long>(duration % 1000000000)};
+      timespec rem;
+      nanosleep(&req, &rem);
+      now = MonotonicClock::now();
+    }
+    return soonest_deadline_idx;
   }
-  fastly_world_list_handle_t hs{.ptr = handles, .len = tasks_len};
+
+  fastly_world_list_handle_t hs{.ptr = handles.data(), .len = handles.size()};
   fastly_world_option_u32_t ret;
   fastly_compute_at_edge_types_error_t err = 0;
-  if (!fastly_compute_at_edge_async_io_select(&hs, 0, &ret, &err)) {
+  if (!fastly_compute_at_edge_async_io_select(&hs, (soonest_deadline - now) / 1000000, &ret,
```

Comment:
It would be nice to factor the constants here out into conversion functions. You could use the `std::chrono::duration` conversion functions, but those get a bit wordy. I'd settle for naming `1000000` as `MS` and `1000000000` as `US` constants somewhere in this file :)

Reply:
Added